File: frontend_v2/src/types/keywords.ts
import type { PaginatedPayload } from './api'

export interface KeywordSummaryItem {
  id: number
  keyword: string
  target_article_id: number | null
  target_article_title: string | null
  target_article_slug: string | null
  search_volume: number
  difficulty: number
  cannibalization_risk: boolean
  status: 'pending' | 'consumed' | string
  created_at: string | null
}

export type KeywordListPayload = PaginatedPayload<KeywordSummaryItem>

export type GapKeywordListPayload = PaginatedPayload<KeywordSummaryItem>

export interface KeywordClusterItem {
  cluster_name: string
  keywords_total: number
  pending_keywords: number
  consumed_keywords: number
  average_difficulty: number | null
}

export interface KeywordClustersPayload {
  items: KeywordClusterItem[]
  limit: number
  warning: string | null
}

export interface KeywordListFilters {
  status?: string
  query?: string
  limit?: number
  offset?: number
}

export interface GapKeywordFilters {
  query?: string
  limit?: number
  offset?: number
}

File: backend/app/schemas/keywords.py
"""Keyword-related schemas."""

from datetime import datetime

from pydantic import BaseModel, Field


class KeywordSummaryItem(BaseModel):
    """Single keyword row for list pages."""

    id: int
    keyword: str
    target_article_id: int | None = None
    target_article_title: str | None = None
    target_article_slug: str | None = None
    search_volume: int = 0
    difficulty: int = 0
    cannibalization_risk: bool = False
    status: str
    created_at: datetime | None = None


class KeywordListPayload(BaseModel):
    """Paginated keyword list response."""

    items: list[KeywordSummaryItem] = Field(default_factory=list)
    total: int
    limit: int
    offset: int
    warning: str | None = None


class KeywordClusterItem(BaseModel):
    """Aggregated keyword cluster item."""

    cluster_name: str
    keywords_total: int
    pending_keywords: int
    consumed_keywords: int
    average_difficulty: float | None = None


class KeywordClustersPayload(BaseModel):
    """Keyword cluster distribution response."""

    items: list[KeywordClusterItem] = Field(default_factory=list)
    limit: int
    warning: str | None = None
